servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -260,15 +248,50 @@ public void onComplete() {
         final ThreadLocalRandom rnd = ThreadLocalRandom.current();
 
         // Try first to see if an existing connection can be used
-        final List<C> connections = host.connections;
-        final int size = connections.size();
-        // With small enough search space, attempt all connections.
-        // Back off after exploring most of the search space, it gives diminishing returns.
-        final int attempts = size < MIN_SEARCH_SPACE ? size : (int) (size * SEARCH_FACTOR);
-        for (int i = 0; i < attempts; i++) {
-            final C connection = connections.get(rnd.nextInt(size));
-            if (selector.test(connection)) {
-                return succeeded(connection);
+        final Object[] connections = host.connections;
+        if (connections.length <= 64) {
+            // With small enough search space, attempt all connections.
+            // Back off after exploring most of the search space, it gives diminishing returns.
+            int i = 0;
+            long indexMask = 0;
+            int collisions = 0;
+            final int collisionThreshold = connections.length >>> 1;
+            while (i < connections.length) {
+                int index = rnd.nextInt(connections.length);
+                long shiftIndex = 1L << index;
+                if ((indexMask & shiftIndex) == 0) {
+                    ++i;
+                    indexMask |= shiftIndex;
+                    @SuppressWarnings("unchecked")
+                    final C connection = (C) connections[index];
+                    if (selector.test(connection)) {
+                        return succeeded(connection);
+                    }
+                } else if (++collisions > collisionThreshold || i > collisionThreshold) {
+                    shiftIndex = 1L;
+                    index = 0;
+                    do {
+                        while ((indexMask & shiftIndex) != 0) {
+                            shiftIndex <<= 1;
+                            ++index;
+                            // no need to set indexMask because we only iterate sequentially at this point.
+                        }
+                        @SuppressWarnings("unchecked")
+                        final C connection = (C) connections[index];
+                        if (selector.test(connection)) {
+                            return succeeded(connection);
+                        }
+                    } while (++i < connections.length);
+                    break;
+                }
+            }
+        } else {
+            for (int i = 0; i < LARGE_LIST_SELECTION_ATTEMPTS; i++) {
```

Comment:
Can we keep the previous logic in this case: `(int) (connections.length * 0.75f)`? 64 attempts looks low if users have 1000s of connections

Reply:
I will remove the selection algorithm part of this PR and move to another PR.